flink-core/src/main/java/org/apache/flink/api/java/ClosureCleaner.java

```diff
@@ -66,14 +67,20 @@ public class ClosureCleaner {
      */
     public static void clean(
             Object func, ExecutionConfig.ClosureCleanerLevel level, boolean checkSerializable) {
-        clean(func, level, checkSerializable, Collections.newSetFromMap(new IdentityHashMap<>()));
+        clean(
+                func,
+                level,
+                checkSerializable,
+                Collections.newSetFromMap(new IdentityHashMap<>()),
+                new Stack<>());
     }
 
     private static void clean(
             Object func,
             ExecutionConfig.ClosureCleanerLevel level,
             boolean checkSerializable,
-            Set<Object> visited) {
+            Set<Object> visited,
+            Stack<Class<?>> clsReferences) {
```

Comment:
Do you see any risk of excessive stack growth in case of complex or deep reference graphs?